wapitiCore/attack/mod_htp.py

```diff
@@ -0,0 +1,183 @@
+            self._db.create_function("REGEXP", 2, regexp)
+
+    async def _analyze_file(self, request: Request):
+        """
+        Retrieves the url's content and then analyze it to get the technology and the version
+        """
+        try:
+            response = await self.crawler.async_send(request, follow_redirects=True)
+        except RequestError:
+            self.network_errors += 1
+            return
+        if response.content is None or len(response.content) == 0:
+            return
+        content_hash = self._analyze_str(response.content)
+        if content_hash is not None:
+            technology = content_hash[0]
+            tech_info = json.loads(content_hash[1])
+
+            if self.tech_versions.get(technology) is None:
+                self.tech_versions[technology] = []
+
+            self.tech_versions[technology].append(json.loads(tech_info)["versions"])
+
+    async def finish(self):
+        if self._db is None:
+            return
+        root_url = await self.persister.get_root_url()
+        truth_table: List[Version] = None
+        ranges_tables = None
+
+        for technology, versions_list in self.tech_versions.items():
+            # First we retrieve all the stored versions in the same order as they have been added to the database
+            truth_table = self._get_versions(technology)
+            ranges_tables = []
+
+            # We create ranges of versions by using the index of the version in the truth table
+            for versions in versions_list:
+                ranges_tables.append([truth_table.index(versions[0]), truth_table.index(versions[len(versions) - 1])])
+
+            # We obtain the list of min range values by only keeping the first value
+            min_range = [arr[0] for arr in ranges_tables]
+
+            # We obtain the list of max range values by only keeping the last value
+            max_range = [arr[len(arr) - 1] for arr in ranges_tables]
+
+            # We get the min range by sorting the ranges by ascending order and retrieving the first value
+            min_index = sorted(min_range)[0]
+
+            # We get the max range by sorting the ranges by descending order and retrieving the first value
+            max_index = sorted(max_range, reverse=True)[0]
+
+            tech_info = {}
+            tech_info["name"] = technology
+            tech_info["versions"] = truth_table[min_index:max_index + 1]
+
+            await self.add_vuln_info(
+                category=WEB_SERVER_VERSIONED,
+                request=Request(root_url),
+                info=json.dumps(tech_info)
+            )
+            log_blue(MSG_TECHNO_VERSIONED, technology, truth_table[min_index], truth_table[max_index])
+        self._db.close()
+        self.finished = True
+
+    def _analyze_str(self, page_content: str) -> Tuple[str, str]:
+        cursor = self._db.cursor()
+        page_hash = hashlib.sha256(page_content.encode("utf-8")).hexdigest()
```

Comment:
You may have some encoding errors here. You should specify the `errors="ignore"` options but IMHO it would be better to directly hash the raw data using `response.bytes` instead of encoding from `response.content` this way you ensure to always have the correct hash even if the server gives an invalid charset